### PR TITLE
Show real posts count from nested categories in Categories component

### PR DESCRIPTION
**Pull request by Rike-cz**
#404 solution tryout

### CODE REVIEW

File: components/Categories.php [hunk 1]
```diff
@@ -76,25 +76,16 @@ public function onRun()
...
-                    ->whereNotNull('rainlab_blog_posts.published_at')
-                    ->where('rainlab_blog_posts.published_at', '<', Carbon::now())
-                    ->whereRaw($prefix.'rainlab_blog_categories.id = '.$prefix.'rainlab_blog_posts_categories.category_id')
-                ;
-            });
-            $categories = $categories->getNested();
-        }
-        else {
-            $categories = BlogCategory::getNested();
+            $iterator = function($categories) use (&$iterator) {
```
Comment (bennothommo) [inline components/Categories.php]: @Rike-cz Please add a space between `function` and the open bracket on this line and the one below.
```suggestion
            $iterator = function ($categories) use (&$iterator) {
```

File: components/Categories.php [hunk 2]
```diff
@@ -76,25 +76,16 @@ public function onRun()
...
+                return $categories->reject(function($category) use (&$iterator) {
+                    if($category->post_count == 0) return true;
```
Comment (bennothommo) [inline components/Categories.php]: @Rike-cz Please use curly braces around all conditional statements, even one-liners, and ensure there is a space between the condition and the brackets as per [PSR-2](https://www.php-fig.org/psr/psr-2/#5-control-structures). Please also do this for the line below as well.
```suggestion
                    if ($category->post_count == 0) {
                        return true;
                    }
```

**Comment by bennothommo** [comment]
@Rike-cz I've tested this and it works, however, it has made the post counts for parent categories appear larger than they are. Each post should be counted exclusively.

For example, if I have three categories:

- Parent
- Child 1 (child of Parent)
- Child 2 (child of Parent)

And I create a post that falls under both the "Child 1" and "Child 2" categories, the post counts come up as follows:

- Parent (2, should be 1)
- Child 1 (1)
- Child 2 (1)

The same thing happens if I put a post into both the parent category and one of the child categories - for example, if I add a post to the "Parent" and "Child 2" categories:

- Parent (2, should be 1)
- Child 1 (0)
- Child 2 (1)

The blog plugin previously only counted posts directly in the category when generating the post count too, so I think a new checkbox property may need to be added to allow users to select whether they want to include posts in subcategories in the post count (defaulting to unticked to maintain backwards compatibility).

Would you be able to have a go at making those changes?

**Comment by Rike-cz** [comment]
You are right. I changed common posts counting behavior to original one. Nested counting is only used in category tree component when displayEmpty is false.

**Comment by bennothommo** [comment]
@Rike-cz great work. Did you have any thoughts about the extra property for showing the nested count in the component markup?

**Comment by Rike-cz** [comment]
@bennothommo I tried something, but still have problems with duplicities (the same post in more siblings or in parent and child) :-( I cannot solve it in simple way (without too many SQL queries).

**Comment by bennothommo** [comment]
@Rike-cz No worries, I'll try and take a look at some point.

**Comment by bennothommo** [comment]
@Rike-cz I'm happy with how this works now, no need for that extra parameter at this stage.

**Pull request merged**